src/Avalonia.Native/ScreenImpl.cs

```diff
@@ -2,66 +2,84 @@
 using System.Collections.Generic;
 using Avalonia.Native.Interop;
 using Avalonia.Platform;
+using MicroCom.Runtime;
+
+#nullable enable
 
 namespace Avalonia.Native
 {
-    internal class ScreenImpl : IScreenImpl, IDisposable
+    internal sealed class AvnScreen(uint displayId)
+        : Screen(new PlatformHandle(new IntPtr(displayId), "CGDirectDisplayID"))
+    {
+        public unsafe void Refresh(IAvnScreens native)
+        {
+            void* localizedName = null;
+            var screen = native.GetScreen(displayId, &localizedName);
+
+            IsPrimary = screen.IsPrimary.FromComBool();
+            Scaling = screen.Scaling;
+            Bounds = screen.Bounds.ToAvaloniaPixelRect();
+            WorkingArea = screen.WorkingArea.ToAvaloniaPixelRect();
+            CurrentOrientation = screen.Orientation switch
+            {
+                AvnScreenOrientation.UnknownOrientation => ScreenOrientation.None,
+                AvnScreenOrientation.Landscape => ScreenOrientation.Landscape,
+                AvnScreenOrientation.Portrait => ScreenOrientation.Portrait,
+                AvnScreenOrientation.LandscapeFlipped => ScreenOrientation.LandscapeFlipped,
+                AvnScreenOrientation.PortraitFlipped => ScreenOrientation.PortraitFlipped,
+                _ => throw new ArgumentOutOfRangeException()
+            };
+
+            using var avnString = MicroComRuntime.CreateProxyOrNullFor<IAvnString>(localizedName, true);
+            DisplayName = avnString?.String;
+        }
+    }
+
+    internal class ScreenImpl : ScreensBaseImpl<uint, AvnScreen>, IDisposable
     {
+        private AvnScreenEvents _events;
         private IAvnScreens _native;
 
-        public ScreenImpl(IAvnScreens native)
+        public ScreenImpl(Func<IAvnScreenEvents, IAvnScreens> factory)
         {
-            _native = native;
+            _events = new AvnScreenEvents(this);
```

Comment:
There is no need to keep a reference to the callback class. Just make sure that the native part keeps a proper ComPtr to CCW and do something like:
```cs
using var events = new AvnScreenEvents(this);
_native = factory(events);
```

That way the callback class will be considered to not be referenced from managed code anymore and CCW will get destroyed when the latest reference dies.

Right now the code isn't safe since CCW lifetime is being controlled solely by the managed Dispose.

Reply:
Done